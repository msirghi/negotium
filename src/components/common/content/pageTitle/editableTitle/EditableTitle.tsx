import { TaskWrapperTitleOptions } from '../../types';
import { FC, useState } from 'react';
import { IconButton, TextField, Tooltip, Typography } from '@mui/material';
import { If } from '../../../utilities/if/If';
import CheckIcon from '@mui/icons-material/Check';
import ClearIcon from '@mui/icons-material/Clear';
import { Row } from '../../../utilities/row/Row';

type Props = {
  title: string;
  editableOptions: TaskWrapperTitleOptions;
  className?: string;
};

export const EditableTitle: FC<Props> = ({ title, editableOptions, className }) => {
  const [fieldValue, setFieldValue] = useState(title);
  const [editMode, setEditMode] = useState(false);
  const [error, setError] = useState('');

  const toggleEditMode = () => setEditMode(!editMode);

  const onSave = () => {
    setError('');
    if (!fieldValue) {
      setError('Field is mandatory.');
      return;
    }
    editableOptions.onSave(fieldValue);
    toggleEditMode();
  };

  const onCancel = () => {
    toggleEditMode();
    setFieldValue(title);
    setError('');
  };

  return (
    <div>
      <If condition={editMode}>
        <Row alignVerticalCenter>
          <TextField
            onChange={(e) => setFieldValue(e.target.value)}
            value={fieldValue}
            variant="standard"
            inputProps={{
              style: { fontSize: 34 },
              'data-testid': 'et-title-field',
            }}
            error={!!error}
            helperText={error}
          />
          <IconButton onClick={onSave} data-testid={'et-save-button'}>
            <Tooltip title={'Save'}>
              <CheckIcon color={'primary'} />
            </Tooltip>
          </IconButton>
          <IconButton onClick={onCancel} data-testid={'et-cancel-button'}>
            <Tooltip title={'Cancel'}>
              <ClearIcon color={'error'} />
            </Tooltip>
          </IconButton>
        </Row>
      </If>

      <If condition={!editMode}>
        <Typography
          onClick={toggleEditMode}
          className={className}
          fontSize={34}
          data-testid="et-title"
        >
          {title}
        </Typography>
      </If>
    </div>
  );
};
