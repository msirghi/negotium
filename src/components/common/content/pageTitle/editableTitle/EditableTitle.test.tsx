import renderer from 'react-test-renderer';
import { EditableTitle } from './EditableTitle';
import { mount } from 'enzyme';
import { TextField, Typography } from '@mui/material';
import { act, fireEvent, render } from '@testing-library/react';

describe('EditableTitle', () => {
  const defaultProps = {
    title: 'Some title',
    editableOptions: {
      title: '',
      onSave: jest.fn(),
    },
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const tree = renderer.create(<EditableTitle {...defaultProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot with provided font size', () => {
    const tree = renderer.create(
      <EditableTitle
        {...defaultProps}
        editableOptions={{
          ...defaultProps.editableOptions,
          inputFontSize: 14,
        }}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it('should show input field on title click', () => {
    const wrapper = mount(<EditableTitle {...defaultProps} />);
    const title = wrapper.find(Typography);

    title.simulate('click');
    wrapper.update();
    expect(wrapper.find(TextField)).toHaveLength(1);
  });

  it('should show an error on submitting an empty value', () => {
    const { getByTestId, getByText } = render(
      <EditableTitle {...defaultProps} />
    );
    const title = getByTestId('et-title');
    act(() => {
      fireEvent.click(title);
    });

    const input = getByTestId('et-title-field');
    act(() => {
      fireEvent.change(input, { target: { value: '' } });
    });

    const submitButton = getByTestId('et-save-button');
    act(() => {
      fireEvent.click(submitButton);
    });
    expect(getByText('Field is mandatory.')).toBeInTheDocument();
  });

  it('should call new method on saving the value', () => {
    const { getByTestId } = render(<EditableTitle {...defaultProps} />);
    const title = getByTestId('et-title');
    act(() => {
      fireEvent.click(title);
    });

    const input = getByTestId('et-title-field');
    act(() => {
      fireEvent.change(input, { target: { value: 'Value' } });
    });

    const submitButton = getByTestId('et-save-button');
    act(() => {
      fireEvent.click(submitButton);
    });
    expect(defaultProps.editableOptions.onSave).toBeCalledWith('Value');
  });

  it('should hide an input if cancel button was clicked', () => {
    const { getByTestId, queryByTestId } = render(
      <EditableTitle {...defaultProps} />
    );
    const title = getByTestId('et-title');
    act(() => {
      fireEvent.click(title);
    });

    const input = getByTestId('et-title-field');
    act(() => {
      fireEvent.change(input, { target: { value: '' } });
    });

    const cancelButton = getByTestId('et-cancel-button');
    act(() => {
      fireEvent.click(cancelButton);
    });
    expect(queryByTestId('et-title-field')).not.toBeInTheDocument();
  });
});
