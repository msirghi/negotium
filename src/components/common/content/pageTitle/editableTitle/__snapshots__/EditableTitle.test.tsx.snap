// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTitle should match the snapshot 1`] = `
<div>
  <p
    className="MuiTypography-root MuiTypography-body1 css-mfnxei-MuiTypography-root"
    data-testid="et-title"
    onClick={[Function]}
  >
    Some title
  </p>
</div>
`;

exports[`EditableTitle should match the snapshot with provided font size 1`] = `
<div>
  <p
    className="MuiTypography-root MuiTypography-body1 css-mfnxei-MuiTypography-root"
    data-testid="et-title"
    onClick={[Function]}
  >
    Some title
  </p>
</div>
`;
